test/e2e/router_status_test.go

```diff
@@ -0,0 +1,297 @@
+		}
+	}()
+
+	// Create a route that will NOT be immediately admitted by the ingress controller, but will be admitted AFTER
+	// the IC selectors are updated. The status SHOULD be successfully admitted.
+	// Use openshift-console namespace to get a namespace outside the ingress-operator's cache.
+	routeBarLabelName := types.NamespacedName{Namespace: "openshift-console", Name: "route-bar-label"}
+	routeBarLabel := newRouteWithLabel(routeBarLabelName, "bar")
+	if err := kclient.Create(context.TODO(), routeBarLabel); err != nil {
+		t.Fatalf("failed to create route: %v", err)
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), routeBarLabel); err != nil {
+			t.Fatalf("failed to delete route %s: %v", routeBarLabel, err)
+		}
+	}()
+
+	// Wait for routeFooLabel to be admitted upon creation.
+	admittedCondition := routev1.RouteIngressCondition{Type: routev1.RouteAdmitted, Status: corev1.ConditionTrue}
+	if err := waitForRouteIngressConditions(t, kclient, routeFooLabelName, ic.Name, admittedCondition); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	// Update the ingress controller to not select routeFooLabel to be admitted, but to select
+	// routeBarLabel instead.
+	if err := kclient.Get(context.TODO(), icName, ic); err != nil {
+		t.Fatalf("failed to get ingress resource: %v", err)
+	}
+	ic.Spec.RouteSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"type": "bar",
+		},
+	}
+	if err := kclient.Update(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to update ingresscontroller: %v", err)
+	}
+
+	// Wait until routeFooLabel status is cleared.
+	if err := waitForRouteStatusClear(t, kclient, routeFooLabelName, ic.Name); err != nil {
+		t.Fatalf("failed to observe route status clear when the ingress controller route selector was updated: %v", err)
+	}
+
+	// Wait until routeBarLabel status is admitted.
+	if err := waitForRouteIngressConditions(t, kclient, routeBarLabelName, ic.Name, admittedCondition); err != nil {
+		t.Fatalf("failed to observe route status admitted when te ingress controller route selector was updated: %v", err)
+	}
+}
+
+func TestIngressControllerNamespaceSelectorUpdateShouldClearRouteStatus(t *testing.T) {
+	t.Parallel()
+	// Create an ingress controller that can admit our route.
+	icName := types.NamespacedName{Namespace: operatorNamespace, Name: "ic-namespace-selector-test"}
+	domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newPrivateController(icName, domain)
+	ic.Spec.NamespaceSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"type": "foo",
+		},
+	}
+
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
+	defer assertIngressControllerDeleted(t, kclient, ic)
+
+	// Create a new namespace for the route that we can immediately match with the IC's namespace selector.
+	nsFoo := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
```

Comment:
This name is rather generic; maybe better to pick a name that is unlikely to be used by something outside of the ingress operator's tests.